tracee-ebpf/tracee/proc.go

```diff
@@ -0,0 +1,87 @@
+package tracee
+
+import (
+	"github.com/prometheus/procfs"
+)
+
+func gatherEntireProcessTree() (map[uint32][]uint32, error) {
+	fs, err := procfs.NewFS("/proc")
+	if err != nil {
+		return nil, err
+	}
+
+	procs, err := fs.AllProcs()
+	if err != nil {
+		return nil, err
+	}
+
+	processMap := map[uint32][]uint32{}
+	for i := range procs {
+		stat, err := procs[i].Stat()
+		if err != nil {
+			return nil, err
+		}
+		processMap[uint32(stat.PID)] = processMap[uint32(stat.PID)]
+		processMap[uint32(stat.PPID)] = append(processMap[uint32(stat.PPID)], uint32(stat.PID))
+	}
+
+	return processMap, nil
+}
+
+// populateProcessTreeFilterMap takes a map of the process tree (k=ppid, v=[]childpid)
+// and a filter specification map (k=pid,v=trace/not) and returns a fully populated map
+// where k=pid, v=trace/not for all pids
+func populateProcessTreeFilterMap(processTree map[uint32][]uint32,
+	filterSpecification map[uint32]bool) map[uint32]bool {
+
+	filterMap := map[uint32]bool{}
+
+	// Determine the default filter for PIDs that aren't specified with a proc tree filter
+	// - If one or more '=' filters, default is '!='
+	// - If one or more '!=' filters, default is '='
+	// - If a mix of filters, the default is '='
+	var defaultFilter = true
+	for _, v := range filterSpecification {
+		defaultFilter = defaultFilter && v
+	}
+	defaultFilter = !defaultFilter
+
+	// Populate inital filter map  (keys representing all pids)
+	for pid, _ := range processTree {
+		filterMap[pid] = defaultFilter
+	}
```

Comment:
I don't think we need to save this value for every pid in the system.
Instead, we can pass the default filter value in the config map, and use it when we don't find a given pid in the map

Reply:
Sure this takes up some extra space, but this adds a few more instructions in each  `should_trace()` call to check the default value if the PID isn't in the map. I don't feel that strongly about this, I can if you'd like.

Reply:
equality_filter_matches (which you use in the bpf code) already takes care of that. You should just set CONFIG_PROC_TREE_FILTER to either FILTER_IN or FILTER_OUT in the config map.

Reply:
That's a go vet error, but i'm still not entirely sure why it doesn't complain in this context.